File: internal/fsapi/mkdir.go
package fsapi

import (
	"encoding/json"
	"errors"
	"io/fs"
	"strings"

	"fsmcp/internal/proto"
)

const mkdirDescription = `Create a directory within the root, recursively, creating any missing parent directories.`

var mkdirInputSchema = json.RawMessage(`{"type":"object","properties":{"path":{"type":"string","description":"Path relative to the root."}},"required":["path"],"additionalProperties":false}`)

// RegisterMkdir wires fs_mkdir into reg.
func RegisterMkdir(reg *Registry) {
	reg.Register(proto.Tool{
		Name:        "fs_mkdir",
		Description: mkdirDescription,
		InputSchema: mkdirInputSchema,
		Annotations: proto.ToolAnnotations{ReadOnlyHint: false, OpenWorldHint: false},
	}, handleMkdir)
}

type mkdirArgs struct {
	Path string `json:"path"`
}

type mkdirResult struct {
	OK      bool     `json:"ok"`
	Path    string   `json:"path"`
	Created []string `json:"created"`
}

// errMkdirBlocked marks a path component that exists but is not a
// directory, so the requested directory could never be created under it.
var errMkdirBlocked = errors.New("path component exists and is not a directory")

func handleMkdir(root *Root, rawArgs json.RawMessage) *proto.CallToolResult {
	var args mkdirArgs
	if err := decodeArgs(rawArgs, &args); err != nil {
		return proto.NewErrorResult(proto.ErrInvalidArgument,
			`arguments must be a JSON object with a "path" string`+": "+err.Error(), "")
	}

	normalized, err := NormalizePath(args.Path)
	if err != nil {
		return Fail(err, args.Path)
	}

	created, mkErr := mkdirAll(root, normalized)
	if mkErr != nil {
		if errors.Is(mkErr, errMkdirBlocked) {
			return proto.NewErrorResult(proto.ErrNotADir, "a path component exists and is not a directory", normalized)
		}
		return Fail(mkErr, args.Path)
	}

	return proto.NewSuccessResult(mkdirResult{OK: true, Path: normalized, Created: created})
}

// mkdirAll creates target and any missing ancestors, walking down from the
// root component by component so each directory this call actually
// created is reported, in the order it came into being. A directory that
// already existed is not repeated in the result.
func mkdirAll(root *Root, target string) ([]string, error) {
	created := []string{}
	if target == "." {
		return created, nil
	}

	parts := strings.Split(target, "/")
	cur := ""
	for _, part := range parts {
		if cur == "" {
			cur = part
		} else {
			cur = cur + "/" + part
		}

		fi, err := root.Lstat(cur)
		switch {
		case err == nil:
			if !fi.IsDir() {
				return created, errMkdirBlocked
			}
		case errors.Is(err, fs.ErrNotExist):
			if err := root.Mkdir(cur, 0o777); err != nil {
				return created, err
			}
			created = append(created, cur)
		default:
			return created, err
		}
	}
	return created, nil
}

File: src/tools/mkdir.ts
import * as fs from 'fs';
import { ToolRegistry, schema, stringProp, boolProp } from '../registry';
import { textResult, errorResult, ToolContext } from '../types';
import { checkPath } from '../security';

export function registerMkdir(registry: ToolRegistry): void {
  registry.register(
    {
      name: 'fs_mkdir',
      description:
        'Create a directory. Creates missing parent directories too unless recursive is set to false.',
      inputSchema: schema(
        {
          path: stringProp('Absolute path of the directory to create'),
          recursive: boolProp('Create missing parent directories (default: true)'),
        },
        ['path']
      ),
      // Creates a local directory only; never contacts anything off this
      // machine.
      annotations: { readOnlyHint: false, openWorldHint: false },
      category: 'File System',
    },
    (args: Record<string, unknown>, ctx: ToolContext) => {
      const dirPath = args.path as string;
      const recursive = (args.recursive as boolean) ?? true;

      const pathErr = checkPath(dirPath, ctx.allowedDirs);
      if (pathErr) return pathErr;

      let existed = false;
      try {
        existed = fs.statSync(dirPath).isDirectory();
      } catch {
        // does not exist yet -- the ordinary case
      }
      if (existed) return textResult(`directory already exists: ${dirPath}`);

      try {
        fs.mkdirSync(dirPath, { recursive });
      } catch (err: unknown) {
        const message = err instanceof Error ? err.message : String(err);
        return errorResult(`mkdir failed: ${message}`);
      }

      return textResult(`Created directory: ${dirPath}`);
    }
  );
}
